File: kmp/desktop/app/src/main/kotlin/dev/mjablonski/rewordme/app/SettingsViewModel.kt
package dev.mjablonski.rewordme.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import dev.mjablonski.rewordme.domain.ApiKeyStore
import dev.mjablonski.rewordme.domain.ConfigStore
import dev.mjablonski.rewordme.domain.ModelSelection
import dev.mjablonski.rewordme.domain.Rewording
import dev.mjablonski.rewordme.data.AccountProviderService
import dev.mjablonski.rewordme.data.AccountProviderStatus
import dev.mjablonski.rewordme.data.LocalModelManager
import dev.mjablonski.rewordme.data.LocalModelProgress
import dev.mjablonski.rewordme.data.LocalModelState
import dev.mjablonski.rewordme.models.HotkeyConfig
import dev.mjablonski.rewordme.models.ModelInfo
import dev.mjablonski.rewordme.models.ProviderKind
import dev.mjablonski.rewordme.models.ProviderAccess
import dev.mjablonski.rewordme.models.LocalModelCatalog
import dev.mjablonski.rewordme.models.OfflineModelManifest
import dev.mjablonski.rewordme.models.RewordConfig
import dev.mjablonski.rewordme.models.RewordError
import dev.mjablonski.rewordme.models.RewriteRule
import dev.mjablonski.rewordme.platform.StartupRegistration
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.awt.Desktop
import java.net.URI
import java.util.Locale

/**
 * Settings state. Every edit is written straight to disk - there is no OK or
 * Cancel button, matching the macOS app.
 */
class SettingsViewModel(
    private val configStore: ConfigStore,
    private val keyStore: ApiKeyStore,
    private val rewordService: Rewording,
    private val accountProviders: AccountProviderService = AccountProviderService(),
    private val localModel: LocalModelManager = LocalModelManager(),
    private val scope: CoroutineScope,
    private val hotkeys: HotkeyBinder
) {
    /**
     * Convenience for the composition root. Tests build the primary
     * constructor with fakes instead, so that nothing has to stand up
     * AppDependencies - which would open the real config file and the real
     * credential store.
     */
    constructor(
        dependencies: AppDependencies,
        scope: CoroutineScope,
        hotkeys: HotkeyBinder
    ) : this(
        dependencies.configStore,
        dependencies.keyStore,
        dependencies.rewordService,
        dependencies.accountProviderService,
        dependencies.localModelManager,
        scope,
        hotkeys
    )

    var config by mutableStateOf(configStore.load())
        private set
    var apiKey by mutableStateOf("")
        private set
    var keySaved by mutableStateOf(false)
        private set
    var keySaveError by mutableStateOf<String?>(null)
        private set

    var availableModels by mutableStateOf(emptyList<ModelInfo>())
        private set
    var isLoadingModels by mutableStateOf(false)
        private set
    var modelsError by mutableStateOf<String?>(null)
        private set

    var accountStatus by mutableStateOf<AccountProviderStatus?>(null)
        private set
    var isCheckingAccount by mutableStateOf(false)
        private set
    var isSigningIn by mutableStateOf(false)
        private set
    var accountError by mutableStateOf<String?>(null)
        private set

    var localModelState by mutableStateOf<LocalModelState>(LocalModelState.NotDownloaded)
        private set
    var localDownloadProgress by mutableStateOf(
        LocalModelProgress(0, LocalModelCatalog.DEFAULT.byteCount)
    )
        private set

    var hotkeyStatus by mutableStateOf(HotkeyStatus.Active)
        private set
    var isRecordingHotkey by mutableStateOf(false)
        private set

    private var modelLoad: Job? = null
    private var modelLoadId = 0L
    private var accountJob: Job? = null
    private var accountRequestId = 0L
    private var localDownloadJob: Job? = null
    private var localDownloadId = 0L
    private var localDownloadModelId: String? = null
    private var persistedApiKey = ""

    init {
        apiKey = keyStore.apiKey(config.provider) ?: ""
        persistedApiKey = apiKey
        refreshProviderSetup()
        if (config.provider.access in setOf(ProviderAccess.ACCOUNT, ProviderAccess.MANAGED_LOCAL)) {
            loadModels()
        }
    }

    val launchAtLoginSupported: Boolean get() = StartupRegistration.isSupported

    var launchAtLogin by mutableStateOf(StartupRegistration.isEnabled)
        private set

    /**
     * Reads the registry back, so the switch snaps to whatever actually stuck,
     * and records the answer so first-run registration never overrides it.
     */
    fun toggleLaunchAtLogin(enabled: Boolean) {
        launchAtLogin = if (StartupRegistration.isSupported) {
            StartupRegistration.isEnabled = enabled
            StartupRegistration.isEnabled
        } else {
            enabled
        }
        update(config.copy(launchAtLogin = launchAtLogin))
    }

    /** The model "Automatic" would pick right now, once the list is known. */
    val automaticModelHint: String?
        get() = ModelSelection.defaultModel(config.provider, availableModels)?.id
            ?.takeUnless { it == "automatic" }

    val canSaveApiKey: Boolean
        get() = apiKey.trim() != persistedApiKey

    fun selectProvider(provider: ProviderKind) {
        if (provider == config.provider) return
        update(config.selectingProvider(provider))
        apiKey = keyStore.apiKey(provider) ?: ""
        persistedApiKey = apiKey
        keySaved = false
        keySaveError = null
        clearModelResults()
        refreshProviderSetup()
        if (provider.access in setOf(ProviderAccess.ACCOUNT, ProviderAccess.MANAGED_LOCAL)) {
            loadModels()
        }
    }

    fun editApiKey(value: String) {
        if (value == apiKey) return
        apiKey = value
        keySaved = false
        keySaveError = null
        clearModelResults()
    }

    fun saveApiKey() {
        val normalized = apiKey.trim()
        val saved = keyStore.setApiKey(config.provider, normalized)
        if (saved) {
            apiKey = normalized
            persistedApiKey = normalized
        }
        keySaved = saved
        keySaveError = if (saved) null else Strings["provider.saveFailed"]
    }

    fun setOllamaHost(host: String) {
        if (host == config.ollamaHost) return
        // A selected model and loaded catalog belong to the old server.
        update(config.selectingModel(null).copy(ollamaHost = host))
        clearModelResults()
    }

    fun selectModel(model: String?) = update(config.selectingModel(model))

    val selectedLocalModel: OfflineModelManifest
        get() = LocalModelCatalog.model(config.selectedModel)

    val isLocalDownloadActive: Boolean
        get() = localDownloadModelId != null

    fun selectLocalModel(modelId: String?) {
        val manifest = LocalModelCatalog.ALL.firstOrNull { it.id == modelId } ?: return
        if (manifest.id == selectedLocalModel.id) return
        update(config.selectingModel(manifest.id))
        refreshLocalModelState()
    }

    fun setBasePrompt(prompt: String) = update(config.copy(basePrompt = prompt))

    fun addRule() =
        update(config.copy(rules = config.rules + RewriteRule(text = "")))

    fun removeRule(id: String) =
        update(config.copy(rules = config.rules.filterNot { it.id == id }))

    fun updateRule(id: String, transform: (RewriteRule) -> RewriteRule) =
        update(config.copy(rules = config.rules.map { if (it.id == id) transform(it) else it }))

    /**
     * The live shortcut is released while recording: the OS would otherwise
     * swallow the very combination the user is trying to re-record.
     */
    fun beginRecording() {
        isRecordingHotkey = true
        hotkeys.release()
    }

    fun cancelRecording() {
        if (!isRecordingHotkey) return
        isRecordingHotkey = false
        hotkeyStatus = hotkeys.bind(config)
    }

    fun applyRecorded(hotkey: HotkeyConfig) {
        isRecordingHotkey = false
        val candidate = config.copy(hotkey = hotkey)
        val status = hotkeys.bind(candidate)
        if (status == HotkeyStatus.Conflict || status == HotkeyStatus.Failed) {
            // Saving a combination we could not claim would leave the user with
            // no working shortcut, so keep the old one and explain.
            hotkeys.bind(config)
            hotkeyStatus = status
        } else {
            update(candidate)
            hotkeyStatus = status
        }
    }

    fun reportHotkeyStatus(status: HotkeyStatus) {
        hotkeyStatus = status
    }

    fun loadModels() {
        modelLoad?.cancel()
        val requestId = ++modelLoadId
        val provider = config.provider
        val key = apiKey
        val endpoint = config.endpointOverride
        isLoadingModels = true
        modelsError = null
        modelLoad = scope.launch {
            try {
                val models = withContext(Dispatchers.IO) {
                    rewordService.listModels(provider, key, endpoint)
                }
                if (requestId != modelLoadId) return@launch
                availableModels = models
                if (models.isEmpty()) modelsError = Strings["provider.noModels"]
            } catch (error: RewordError) {
                if (requestId == modelLoadId) modelsError = error.localized()
            } catch (error: Exception) {
                if (error is kotlinx.coroutines.CancellationException) throw error
                if (requestId == modelLoadId) {
                    modelsError = error.message ?: Strings["provider.noModels"]
                }
            } finally {
                if (requestId == modelLoadId) isLoadingModels = false
            }
        }
    }

    fun refreshProviderSetup() {
        accountJob?.cancel()
        accountJob = null
        val requestId = ++accountRequestId
        accountStatus = null
        accountError = null
        isCheckingAccount = false
        isSigningIn = false
        val provider = config.provider
        when (provider.access) {
            ProviderAccess.ACCOUNT -> {
                isCheckingAccount = true
                accountJob = scope.launch {
                    try {
                        val status = withContext(Dispatchers.IO) { accountProviders.status(provider) }
                        if (requestId != accountRequestId || config.provider != provider) return@launch
                        accountStatus = status
                    } catch (error: Exception) {
                        if (error is kotlinx.coroutines.CancellationException) throw error
                        if (requestId == accountRequestId && config.provider == provider) {
                            accountError = userMessage(error)
                        }
                    } finally {
                        if (requestId == accountRequestId) {
                            isCheckingAccount = false
                            isSigningIn = false
                            accountJob = null
                        }
                    }
                }
            }
            ProviderAccess.MANAGED_LOCAL -> refreshLocalModelState()
            ProviderAccess.API_KEY, ProviderAccess.EXTERNAL_LOCAL -> Unit
        }
    }

    fun setUpAccountProvider() {
        val provider = config.provider
        if (!provider.isAccountProvider) return
        accountJob?.cancel()
        val requestId = ++accountRequestId
        isCheckingAccount = true
        isSigningIn = false
        accountError = null
        accountJob = scope.launch {
            try {
                val status = withContext(Dispatchers.IO) { accountProviders.status(provider) }
                if (requestId != accountRequestId || config.provider != provider) return@launch
                accountStatus = status
                isCheckingAccount = false
                if (!status.isInstalled) {
                    runCatching { Desktop.getDesktop().browse(URI(provider.apiKeyConsoleUrl)) }
                        .onFailure {
                            if (requestId == accountRequestId) accountError = userMessage(it)
                        }
                    return@launch
                }
                isSigningIn = true
                withContext(Dispatchers.IO) { accountProviders.signIn(provider) }
                if (requestId != accountRequestId || config.provider != provider) return@launch
                val refreshed = withContext(Dispatchers.IO) { accountProviders.status(provider) }
                if (requestId != accountRequestId || config.provider != provider) return@launch
                accountStatus = refreshed
            } catch (error: Exception) {
                if (error is kotlinx.coroutines.CancellationException) throw error
                if (requestId == accountRequestId && config.provider == provider) {
                    accountError = userMessage(error)
                }
            } finally {
                if (requestId == accountRequestId) {
                    isCheckingAccount = false
                    isSigningIn = false
                    accountJob = null
                }
            }
        }
    }

    fun downloadLocalModel() {
        if (localDownloadJob != null) return
        val manifest = selectedLocalModel
        val requestId = ++localDownloadId
        localDownloadModelId = manifest.id
        val initial = LocalModelProgress(0, manifest.byteCount)
        localDownloadProgress = initial
        localModelState = LocalModelState.Downloading(initial)
        localDownloadJob = scope.launch {
            try {
                withContext(Dispatchers.IO) {
                    localModel.download(manifest.id) { progress ->
                        scope.launch {
                            if (requestId == localDownloadId) {
                                localDownloadProgress = progress
                                localModelState = LocalModelState.Downloading(progress)
                            }
                        }
                    }
                }
                val state = withContext(Dispatchers.IO) { localModel.state(manifest.id) }
                if (requestId == localDownloadId) localModelState = state
            } catch (error: Exception) {
                if (error is kotlinx.coroutines.CancellationException) {
                    if (requestId == localDownloadId) {
                        localModelState = LocalModelState.NotDownloaded
                    }
                    throw error
                }
                if (requestId == localDownloadId) {
                    localModelState = LocalModelState.Failed(
                        userMessage(error)
                    )
                }
            } finally {
                if (localDownloadModelId == manifest.id) localDownloadModelId = null
                localDownloadJob = null
            }
        }
    }

    fun cancelLocalModelDownload() {
        val job = localDownloadJob ?: return
        ++localDownloadId
        localDownloadModelId = null
        localModelState = LocalModelState.NotDownloaded
        // Mark the coroutine cancelled before closing its blocking stream so
        // the resulting IOException is correctly treated as cancellation.
        job.cancel()
        localModel.cancelDownload()
    }

    fun removeLocalModel() {
        val manifest = selectedLocalModel
        cancelLocalModelDownload()
        scope.launch {
            try {
                withContext(Dispatchers.IO) { localModel.removeModel(manifest.id) }
                if (selectedLocalModel.id == manifest.id) {
                    localModelState = LocalModelState.NotDownloaded
                }
            } catch (error: Exception) {
                if (selectedLocalModel.id == manifest.id) {
                    localModelState = LocalModelState.Failed(userMessage(error))
                }
            }
        }
    }

    val localProgressText: String
        get() = "${formatBytes(localDownloadProgress.receivedBytes)} / ${formatBytes(localDownloadProgress.totalBytes)}"

    fun formatBytes(bytes: Long): String = when {
        bytes >= 1_000_000_000 -> String.format(Locale.getDefault(), "%.2f GB", bytes / 1_000_000_000.0)
        bytes >= 1_000_000 -> String.format(Locale.getDefault(), "%.1f MB", bytes / 1_000_000.0)
        bytes >= 1_000 -> String.format(Locale.getDefault(), "%.1f KB", bytes / 1_000.0)
        else -> "$bytes B"
    }

    private fun refreshLocalModelState() {
        val provider = config.provider
        val modelId = selectedLocalModel.id
        scope.launch {
            val state = withContext(Dispatchers.IO) { localModel.state(modelId) }
            if (config.provider == provider && selectedLocalModel.id == modelId) {
                localModelState = state
            }
        }
    }

    private fun clearModelResults() {
        modelLoad?.cancel()
        modelLoad = null
        modelLoadId++
        isLoadingModels = false
        availableModels = emptyList()
        modelsError = null
    }

    private fun userMessage(error: Throwable): String = when (error) {
        is RewordError -> error.localized()
        else -> error.message ?: error.javaClass.simpleName
    }

    private fun update(updated: RewordConfig) {
        config = updated
        configStore.save(updated)
    }
}

enum class HotkeyStatus { Active, Conflict, Failed, Unsupported }

/** Lets the settings screen rebind the global shortcut without owning it. */
interface HotkeyBinder {
    fun bind(config: RewordConfig): HotkeyStatus
    fun release()
}

File: macos/Sources/RewordMeApp/SettingsViewModel.swift
import AppKit
import Foundation
import RewordMeData
import RewordMeDomain
import RewordMeModels
import RewordMePlatform
import ServiceManagement
import SwiftUI

extension Notification.Name {
    /// Posted after every config save so the app delegate can re-apply
    /// behavior that lives outside the Settings window (the hotkey).
    static let rewordConfigChanged = Notification.Name("RewordMeConfigChanged")
}

/// View model for the Settings window. Config changes save to disk
/// immediately; API keys go straight to the injected key store.
@MainActor
final class SettingsViewModel: ObservableObject {
    @Published var config: RewordConfig {
        didSet {
            try? configStore.save(config)
            NotificationCenter.default.post(name: .rewordConfigChanged, object: nil)
            if oldValue.provider != config.provider {
                apiKey = keyStore.apiKey(for: config.provider) ?? ""
                persistedAPIKey = apiKey
                keySavedFeedback = false
                keySaveError = nil
                clearModelResults()
                refreshProviderSetup()
                if config.provider.access == .account || config.provider.access == .managedLocal {
                    loadModels()
                }
            }
            if oldValue.ollamaHost != config.ollamaHost {
                invalidateResolvedModel()
            }
            if oldValue.provider == .local,
               config.provider == .local,
               oldValue.model != config.model {
                refreshLocalModelState()
                invalidateResolvedModel()
            }
        }
    }

    @Published var apiKey: String = ""
    @Published var availableModels: [ModelInfo] = []
    @Published var isLoadingModels = false
    @Published var modelsError: String?
    @Published var keySavedFeedback = false
    @Published var keySaveError: String?
    @Published var accountStatus: AccountProviderStatus?
    @Published var isCheckingAccount = false
    @Published var isSigningIn = false
    @Published var accountError: String?
    @Published var localModelState: LocalModelState = .notDownloaded
    @Published var localDownloadProgress = LocalModelProgress(
        receivedBytes: 0,
        totalBytes: LocalModelManifest.byteCount
    )
    @Published var accessibilityTrusted: Bool
    @Published var launchAtLogin: Bool {
        didSet { updateLaunchAtLogin() }
    }

    private let configStore: ConfigStore
    private let keyStore: any APIKeyStore
    private let service: RewordService
    private let modelResolver: ModelResolver
    private let accountProviders: AccountProviderService
    private let localModel: LocalModelManager
    private let accessibility: any AccessibilityChecking
    private let defaults: UserDefaults
    private var feedbackTask: Task<Void, Never>?
    private var modelLoadTask: Task<Void, Never>?
    private var modelLoadID: UUID?
    private var accountTask: Task<Void, Never>?
    private var accountRequestID: UUID?
    private var localDownloadTask: Task<Void, Never>?
    private var localDownloadID: UUID?
    private var localDownloadModelID: String?
    private var persistedAPIKey = ""

    init(dependencies: AppDependencies, defaults: UserDefaults = .standard) {
        self.configStore = dependencies.configStore
        self.keyStore = dependencies.keyStore
        self.service = dependencies.rewordService
        self.modelResolver = dependencies.modelResolver
        self.accountProviders = dependencies.accountProviderService
        self.localModel = dependencies.localModelManager
        self.accessibility = dependencies.accessibility
        self.defaults = defaults

        let loaded = dependencies.configStore.load()
        config = loaded
        launchAtLogin = SMAppService.mainApp.status == .enabled
        accessibilityTrusted = dependencies.accessibility.isTrusted
        apiKey = keyStore.apiKey(for: loaded.provider) ?? ""
        persistedAPIKey = apiKey

        Task { [weak self] in
            self?.refreshProviderSetup()
            if loaded.provider.access == .account || loaded.provider.access == .managedLocal {
                self?.loadModels()
            }
        }
    }

    func saveAPIKey() {
        explainKeychainPromptOnce()
        let normalized = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        let saved = keyStore.setAPIKey(normalized, for: config.provider)
        if saved {
            apiKey = normalized
            persistedAPIKey = normalized
        }
        keySavedFeedback = saved
        keySaveError = saved ? nil : Loc.saveKeyFailed
        guard saved else {
            feedbackTask?.cancel()
            return
        }

        invalidateResolvedModel()
        feedbackTask?.cancel()
        feedbackTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(2.5))
            guard !Task.isCancelled else { return }
            self?.keySavedFeedback = false
        }
    }

    func selectProvider(_ provider: ProviderKind) {
        guard provider != config.provider else { return }
        var updated = config
        updated.provider = provider
        config = updated
    }

    var selectedLocalModel: OfflineModelManifest {
        LocalModelCatalog.model(id: config.model)
    }

    var isLocalDownloadActive: Bool { localDownloadModelID != nil }

    func selectLocalModel(_ modelID: String) {
        guard LocalModelCatalog.all.contains(where: { $0.id == modelID }) else { return }
        var updated = config
        updated.model = modelID
        config = updated
    }

    func editAPIKey(_ value: String) {
        guard value != apiKey else { return }
        apiKey = value
        keySavedFeedback = false
        keySaveError = nil
        clearModelResults()
    }

    var canSaveAPIKey: Bool {
        apiKey.trimmingCharacters(in: .whitespacesAndNewlines) != persistedAPIKey
    }

    func setOllamaHost(_ host: String) {
        guard host != config.ollamaHost else { return }
        var updated = config
        updated.ollamaHost = host
        // A pinned model belongs to the old local server's catalog.
        updated.model = nil
        config = updated
    }

    func loadModels() {
        modelLoadTask?.cancel()
        let requestID = UUID()
        modelLoadID = requestID
        isLoadingModels = true
        modelsError = nil
        let provider = config.provider
        let key = apiKey
        let endpoint = config.endpointOverride
        modelLoadTask = Task { [weak self] in
            guard let self else { return }
            defer {
                if self.modelLoadID == requestID {
                    self.isLoadingModels = false
                    self.modelLoadTask = nil
                }
            }
            do {
                let models = try await service.listModels(
                    provider: provider, apiKey: key, endpoint: endpoint
                )
                guard !Task.isCancelled,
                      self.modelLoadID == requestID,
                      self.config.provider == provider,
                      self.apiKey == key,
                      self.config.endpointOverride == endpoint else { return }
                // Preserve server order: Ollama's first item is its automatic
                // choice. The Picker sorts only its visual presentation.
                self.availableModels = models
                if models.isEmpty {
                    self.modelsError = Loc.noModels
                }
            } catch is CancellationError {
                return
            } catch {
                guard self.modelLoadID == requestID else { return }
                self.modelsError = (error as? RewordError).map(Loc.message(for:))
                    ?? error.localizedDescription
            }
        }
    }

    // MARK: - Account and managed-local setup

    func refreshProviderSetup() {
        accountTask?.cancel()
        accountTask = nil
        accountRequestID = nil
        accountStatus = nil
        accountError = nil
        isCheckingAccount = false
        isSigningIn = false

        let provider = config.provider
        switch provider.access {
        case .account:
            let requestID = UUID()
            accountRequestID = requestID
            isCheckingAccount = true
            accountTask = Task { [weak self] in
                guard let self else { return }
                defer {
                    if self.accountRequestID == requestID {
                        self.isCheckingAccount = false
                        self.isSigningIn = false
                        self.accountTask = nil
                        self.accountRequestID = nil
                    }
                }
                let status = await accountProviders.status(for: provider)
                guard !Task.isCancelled,
                      self.accountRequestID == requestID,
                      self.config.provider == provider else { return }
                self.accountStatus = status
            }
        case .managedLocal:
            refreshLocalModelState()
        case .apiKey, .externalLocal:
            break
        }
    }

    /// Installed CLIs open their official browser sign-in. Missing CLIs open
    /// the provider's official setup page instead of failing silently.
    func setUpAccountProvider() {
        guard config.provider.isAccountProvider else { return }
        accountTask?.cancel()
        let provider = config.provider
        let requestID = UUID()
        accountRequestID = requestID
        isCheckingAccount = true
        isSigningIn = false
        accountError = nil
        accountTask = Task { [weak self] in
            guard let self else { return }
            defer {
                if self.accountRequestID == requestID {
                    self.isCheckingAccount = false
                    self.isSigningIn = false
                    self.accountTask = nil
                    self.accountRequestID = nil
                }
            }
            let status = await accountProviders.status(for: provider)
            guard !Task.isCancelled,
                  self.accountRequestID == requestID,
                  self.config.provider == provider else { return }
            self.accountStatus = status
            self.isCheckingAccount = false
            guard status.isInstalled else {
                NSWorkspace.shared.open(provider.apiKeyConsoleURL)
                return
            }
            self.isSigningIn = true
            do {
                try await accountProviders.signIn(to: provider)
                guard !Task.isCancelled,
                      self.accountRequestID == requestID,
                      self.config.provider == provider else { return }
                let refreshed = await accountProviders.status(for: provider)
                guard !Task.isCancelled,
                      self.accountRequestID == requestID,
                      self.config.provider == provider else { return }
                self.accountStatus = refreshed
            } catch is CancellationError {
                return
            } catch {
                guard self.accountRequestID == requestID,
                      self.config.provider == provider else { return }
                self.accountError = self.localizedMessage(for: error)
            }
        }
    }

    func downloadLocalModel() {
        guard localDownloadTask == nil else { return }
        let manifest = selectedLocalModel
        let downloadID = UUID()
        localDownloadID = downloadID
        localDownloadModelID = manifest.id
        localDownloadProgress = LocalModelProgress(
            receivedBytes: 0,
            totalBytes: manifest.byteCount
        )
        localModelState = .downloading(localDownloadProgress)
        localDownloadTask = Task { [weak self] in
            guard let self else { return }
            do {
                try await localModel.download(modelID: manifest.id) { [weak self] progress in
                    Task { @MainActor in
                        guard let self else { return }
                        guard self.localDownloadID == downloadID else { return }
                        self.localDownloadProgress = progress
                        self.localModelState = .downloading(progress)
                    }
                }
                let state = await localModel.state(modelID: manifest.id)
                if self.localDownloadID == downloadID {
                    self.localModelState = state
                }
            } catch is CancellationError {
                if self.localDownloadID == downloadID {
                    self.localModelState = .notDownloaded
                }
            } catch {
                if self.localDownloadID == downloadID {
                    self.localModelState = .failed(self.localizedMessage(for: error))
                }
            }
            if self.localDownloadID == downloadID { self.localDownloadID = nil }
            if self.localDownloadModelID == manifest.id { self.localDownloadModelID = nil }
            self.localDownloadTask = nil
        }
    }

    func cancelLocalModelDownload() {
        guard localDownloadTask != nil else { return }
        // Invalidate already queued progress callbacks before signalling the
        // downloader so Cancel cannot be followed by stale progress UI.
        localDownloadID = nil
        localDownloadModelID = nil
        localModelState = .notDownloaded
        localDownloadTask?.cancel()
        Task { [localModel] in await localModel.cancelDownload() }
    }

    func removeLocalModel() {
        let manifest = selectedLocalModel
        localDownloadTask?.cancel()
        localDownloadTask = nil
        localDownloadID = nil
        Task { [weak self] in
            guard let self else { return }
            do {
                try await localModel.removeModel(modelID: manifest.id)
                if selectedLocalModel.id == manifest.id { localModelState = .notDownloaded }
            } catch {
                if selectedLocalModel.id == manifest.id {
                    localModelState = .failed(localizedMessage(for: error))
                }
            }
        }
    }

    private func refreshLocalModelState() {
        let provider = config.provider
        let modelID = selectedLocalModel.id
        Task { [weak self] in
            guard let self else { return }
            let state = await localModel.state(modelID: modelID)
            guard self.config.provider == provider,
                  self.selectedLocalModel.id == modelID else { return }
            self.localModelState = state
        }
    }

    var localProgressText: String {
        let formatter = ByteCountFormatter()
        formatter.countStyle = .file
        let received = formatter.string(fromByteCount: localDownloadProgress.receivedBytes)
        let total = formatter.string(fromByteCount: localDownloadProgress.totalBytes)
        return "\(received) / \(total)"
    }

    func formattedBytes(_ bytes: Int64) -> String {
        let formatter = ByteCountFormatter()
        formatter.countStyle = .file
        return formatter.string(fromByteCount: bytes)
    }

    var automaticModelHint: String {
        guard !availableModels.isEmpty else { return "" }
        guard let pick = ModelSelection.defaultModel(for: config.provider, from: availableModels) else {
            return ""
        }
        // Account CLIs expose an internal "automatic" sentinel. The picker
        // already has a localized Automatic row, so repeating the sentinel as
        // a resolved model would be both redundant and misleading.
        guard pick.id != "automatic" else { return "" }
        return Loc.automaticHint(pick.id)
    }

    // MARK: - Rules

    func addRule() {
        config.rules.append(RewriteRule(text: ""))
    }

    func rule(id: UUID) -> RewriteRule? {
        config.rules.first { $0.id == id }
    }

    func updateRule(id: UUID, _ update: (inout RewriteRule) -> Void) {
        guard let index = config.rules.firstIndex(where: { $0.id == id }) else { return }
        var rule = config.rules[index]
        update(&rule)
        config.rules[index] = rule
    }

    func removeRule(id: UUID) {
        config.rules.removeAll { $0.id == id }
    }

    // MARK: - Accessibility

    func refreshAccessibilityStatus() {
        accessibilityTrusted = accessibility.isTrusted
    }

    func openAccessibilitySettings() {
        accessibility.openSystemSettings()
    }

    // MARK: - Private

    private func invalidateResolvedModel() {
        Task { [modelResolver] in await modelResolver.invalidate() }
        clearModelResults()
    }

    private func clearModelResults() {
        modelLoadTask?.cancel()
        modelLoadTask = nil
        modelLoadID = nil
        isLoadingModels = false
        availableModels = []
        modelsError = nil
    }

    private func localizedMessage(for error: Error) -> String {
        (error as? RewordError).map(Loc.message(for:)) ?? error.localizedDescription
    }

    /// Shown once, before the first key ever lands in the Keychain, so the
    /// system's later "RewordMe wants to access your keychain" password
    /// prompt is expected instead of alarming.
    private func explainKeychainPromptOnce() {
        let flag = "keychainPromptExplained"
        guard !defaults.bool(forKey: flag) else { return }
        defaults.set(true, forKey: flag)

        let alert = NSAlert()
        alert.messageText = Loc.keychainAlertTitle
        alert.informativeText = Loc.keychainAlertBody
        alert.alertStyle = .informational
        alert.addButton(withTitle: Loc.gotIt)
        alert.runModal()
    }

    private func updateLaunchAtLogin() {
        do {
            if launchAtLogin {
                try SMAppService.mainApp.register()
            } else {
                try SMAppService.mainApp.unregister()
            }
        } catch {
            launchAtLogin = SMAppService.mainApp.status == .enabled
        }
        // Recorded so first-run registration never overrides an explicit "off".
        config.launchAtLogin = launchAtLogin
    }
}
